File: src/client/java/com/somefrills/features/misc/Freecam.java
package com.somefrills.features.misc;

import com.somefrills.config.FrillsConfig;
import com.somefrills.config.misc.FreecamConfig;
import com.somefrills.events.*;
import com.somefrills.features.core.ToggleFeature;
import com.somefrills.misc.Input;
import com.somefrills.misc.KeyAction;
import com.somefrills.misc.Utils;
import meteordevelopment.orbit.EventHandler;
import meteordevelopment.orbit.EventPriority;
import net.minecraft.client.option.Perspective;
import net.minecraft.entity.Entity;
import net.minecraft.network.packet.s2c.play.DeathMessageS2CPacket;
import net.minecraft.network.packet.s2c.play.HealthUpdateS2CPacket;
import net.minecraft.network.packet.s2c.play.PlayerRespawnS2CPacket;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;
import org.joml.Vector3d;
import org.lwjgl.glfw.GLFW;

import static com.somefrills.Main.mc;
import static com.somefrills.misc.Utils.info;

public class Freecam extends ToggleFeature {
    private final FreecamConfig config;

    public final Vector3d pos = new Vector3d();
    public final Vector3d prevPos = new Vector3d();

    private Perspective perspective = null;

    public float yaw, pitch;
    public float lastYaw, lastPitch;

    private double fovScale;
    private boolean bobView;

    private boolean forward, backward, right, left, up, down;

    public Freecam() {
        super(FrillsConfig.instance.misc.freecam.enabled, FrillsConfig.instance.misc.freecam.keybind);
        config = FrillsConfig.instance.misc.freecam;
    }

    @Override
    public void onEnable() {
        super.onEnable();
        if (mc.options == null || mc.player == null || mc.gameRenderer == null || mc.gameRenderer.getCamera() == null) {
            return;
        }

        fovScale = mc.options.getFovEffectScale().getValue();
        bobView = mc.options.getBobView().getValue();
        if (config.staticView) {
            mc.options.getFovEffectScale().setValue((double) 0);
            mc.options.getBobView().setValue(false);
        }
        yaw = mc.player.getYaw();
        pitch = mc.player.getPitch();

        perspective = mc.options.getPerspective();
        config.speed = config.speed;

        Utils.set(pos, mc.gameRenderer.getCamera().getCameraPos());
        Utils.set(prevPos, mc.gameRenderer.getCamera().getCameraPos());

        if (mc.options.getPerspective() == Perspective.THIRD_PERSON_FRONT) {
            yaw += 180;
            pitch *= -1;
        }

        lastYaw = yaw;
        lastPitch = pitch;

        //isSneaking = mc.options.sneakKey.isPressed();

        forward = Input.isPressed(mc.options.forwardKey);
        backward = Input.isPressed(mc.options.backKey);
        right = Input.isPressed(mc.options.rightKey);
        left = Input.isPressed(mc.options.leftKey);
        up = Input.isPressed(mc.options.jumpKey);
        down = Input.isPressed(mc.options.sneakKey);

        unpress();
        if (config.reloadChunks) mc.worldRenderer.reload();
    }

    @Override
    public void onDisable() {
        super.onDisable();
        if (perspective == null) return;
        if (config.reloadChunks) {
            mc.execute(mc.worldRenderer::reload);
        }

        mc.options.setPerspective(perspective);

        if (config.staticView) {
            mc.options.getFovEffectScale().setValue(fovScale);
            mc.options.getBobView().setValue(bobView);
        }

        perspective = null;
        //isSneaking = false;
    }

    private void unpress() {
        mc.options.forwardKey.setPressed(false);
        mc.options.backKey.setPressed(false);
        mc.options.rightKey.setPressed(false);
        mc.options.leftKey.setPressed(false);
        mc.options.jumpKey.setPressed(false);
        mc.options.sneakKey.setPressed(false);
    }

    @EventHandler
    private void onOpenScreen(ScreenOpenEvent event) {
        unpress();

        prevPos.set(pos);
        lastYaw = yaw;
        lastPitch = pitch;
    }

    @EventHandler
    private void onTick(TickEventPost event) {
        if (mc.getCameraEntity() == null) return;
        if (mc.getCameraEntity().isInsideWall()) mc.getCameraEntity().noClip = true;
        if (!perspective.isFirstPerson()) mc.options.setPerspective(Perspective.FIRST_PERSON);

        Vec3d forward = Vec3d.fromPolar(0, yaw);
        Vec3d right = Vec3d.fromPolar(0, yaw + 90);
        double velX = 0;
        double velY = 0;
        double velZ = 0;

        double s = 0.5;
        if (Input.isPressed(mc.options.sprintKey)) s = 1;

        boolean a = false;
        if (this.forward) {
            velX += forward.x * s * config.speed;
            velZ += forward.z * s * config.speed;
            a = true;
        }
        if (this.backward) {
            velX -= forward.x * s * config.speed;
            velZ -= forward.z * s * config.speed;
            a = true;
        }

        boolean b = false;
        if (this.right) {
            velX += right.x * s * config.speed;
            velZ += right.z * s * config.speed;
            b = true;
        }
        if (this.left) {
            velX -= right.x * s * config.speed;
            velZ -= right.z * s * config.speed;
            b = true;
        }

        if (a && b) {
            double diagonal = 1 / Math.sqrt(2);
            velX *= diagonal;
            velZ *= diagonal;
        }

        if (this.up) {
            velY += s * config.speed;
        }
        if (this.down) {
            velY -= s * config.speed;
        }

        prevPos.set(pos);
        pos.set(pos.x + velX, pos.y + velY, pos.z + velZ);
    }

    @EventHandler(priority = EventPriority.HIGH)
    public void onKey(InputEvent event) {
        if (Input.isKeyPressed(GLFW.GLFW_KEY_F3)) return;

        if (onInput(event.key, event.action)) event.cancel();
    }

    @EventHandler(priority = EventPriority.HIGH)
    private void onMouseClick(MouseClickEvent event) {
        if (onInput(event.button(), event.action)) event.cancel();
    }

    private boolean onInput(int key, KeyAction action) {
        if (Input.getKey(mc.options.forwardKey) == key) {
            forward = action != KeyAction.Release;
            mc.options.forwardKey.setPressed(false);
        } else if (Input.getKey(mc.options.backKey) == key) {
            backward = action != KeyAction.Release;
            mc.options.backKey.setPressed(false);
        } else if (Input.getKey(mc.options.rightKey) == key) {
            right = action != KeyAction.Release;
            mc.options.rightKey.setPressed(false);
        } else if (Input.getKey(mc.options.leftKey) == key) {
            left = action != KeyAction.Release;
            mc.options.leftKey.setPressed(false);
        } else if (Input.getKey(mc.options.jumpKey) == key) {
            up = action != KeyAction.Release;
            mc.options.jumpKey.setPressed(false);
        } else if (Input.getKey(mc.options.sneakKey) == key) {
            down = action != KeyAction.Release;
            mc.options.sneakKey.setPressed(false);
        } else {
            return false;
        }

        return true;
    }

    @EventHandler(priority = EventPriority.LOW)
    private void onMouseScroll(MouseScrollEvent event) {
        if (config.speedScrollSensitivity > 0 && mc.currentScreen == null) {
            config.speed += event.value * 0.25 * (config.speedScrollSensitivity * config.speed);
            if (config.speed < 0.1) config.speed = 0.1;

            event.cancel();
        }
    }


    @EventHandler
    private void onChunkOcclusion(ChunkOcclusionEvent event) {
        event.cancel();
    }

    @EventHandler
    private void onGameLeft(GameStopEvent event) {
        toggle();
    }

    @EventHandler
    private void onPacketReceive(ReceivePacketEvent event) {
        if (event.packet instanceof DeathMessageS2CPacket packet) {
            Entity entity = mc.world.getEntityById(packet.playerId());
            if (entity == mc.player && config.toggleOnDeath) {
                toggle();
                info("Toggled off because you died.");
            }
        } else if (event.packet instanceof HealthUpdateS2CPacket packet) {
            if (mc.player.getHealth() - packet.getHealth() > 0 && config.toggleOnDamage) {
                toggle();
                info("Toggled off because you took damage.");
            }
        } else if (event.packet instanceof PlayerRespawnS2CPacket) {
            toggle();
            info("Toggled off because you changed dimensions.");
        }
    }

    public double getX(float tickDelta) {
        return MathHelper.lerp(tickDelta, prevPos.x, pos.x);
    }

    public double getY(float tickDelta) {
        return MathHelper.lerp(tickDelta, prevPos.y, pos.y);
    }

    public double getZ(float tickDelta) {
        return MathHelper.lerp(tickDelta, prevPos.z, pos.z);
    }

    public double getYaw(float tickDelta) {
        return MathHelper.lerp(tickDelta, lastYaw, yaw);
    }

    public double getPitch(float tickDelta) {
        return MathHelper.lerp(tickDelta, lastPitch, pitch);
    }

    public boolean shouldRenderHands() {
        return config.renderHands;
    }

    public void changeLookDirection(double deltaX, double deltaY) {
        lastYaw = yaw;
        lastPitch = pitch;

        yaw += (float) deltaX;
        pitch += (float) deltaY;

        pitch = MathHelper.clamp(pitch, -90, 90);
    }
}

File: src/main/kotlin/com/somefrills/features/misc/Freecam.kt
package com.somefrills.features.misc

import com.somefrills.Main
import com.somefrills.Main.mc
import com.somefrills.config.FrillsMod
import com.somefrills.events.*
import com.somefrills.modules.FrillsFeature
import com.somefrills.features.core.ToggleFeature
import com.somefrills.misc.KeyAction
import com.somefrills.mixin.KeyMappingAccessor
import com.somefrills.utils.ChatUtils
import com.somefrills.utils.set
import com.somefrills.events.core.EventHandle
import com.somefrills.events.core.EventPriority
import net.minecraft.client.CameraType
import net.minecraft.client.KeyMapping
import net.minecraft.network.protocol.game.ClientboundPlayerCombatKillPacket
import net.minecraft.network.protocol.game.ClientboundRespawnPacket
import net.minecraft.network.protocol.game.ClientboundSetHealthPacket
import net.minecraft.util.Mth
import net.minecraft.world.phys.Vec3
import org.joml.Vector3d
import org.lwjgl.glfw.GLFW
import kotlin.math.sqrt


@FrillsFeature
object Freecam : ToggleFeature(FrillsMod.config.misc.freecam.enabled, FrillsMod.config.misc.freecam.keybind) {
    private val config get() = FrillsMod.config.misc.freecam

    @JvmField
    val pos: Vector3d = Vector3d()

    @JvmField
    val prevPos: Vector3d = Vector3d()

    private var perspective: CameraType? = null

    @JvmField
    var yaw: Float = 0f

    @JvmField
    var pitch: Float = 0f

    @JvmField
    var lastYaw: Float = 0f

    @JvmField
    var lastPitch: Float = 0f

    private var fovScale = 1.0
    private var bobView = false

    private var forward = false
    private var backward = false
    private var right = false
    private var left = false
    private var up = false
    private var down = false

    public override fun onActivate() {
        super.onActivate()
        val player = mc.player ?: return
        fovScale = mc.options.fovEffectScale.get()
        bobView = mc.options.bobView.get()
        if (config.staticView) {
            mc.options.fovEffectScale.set(0.0)
            mc.options.bobView.set(false)
        }

        yaw = player.getYRot();
        pitch = player.getXRot();

        perspective = mc.options.cameraType

        // FIXME: why is this here?
        //config.speed = config.speed

        pos.set(mc.gameRenderer.mainCamera.position())
        prevPos.set(mc.gameRenderer.mainCamera.position())

        if (mc.options.cameraType == CameraType.THIRD_PERSON_FRONT) {
            yaw += 180f
            pitch *= -1f
        }

        lastYaw = yaw
        lastPitch = pitch

        // isSneaking = mc.options.sneakKey.isPressed;
        forward = mc.options.keyUp.isDown
        backward = mc.options.keyDown.isDown
        right = mc.options.keyRight.isDown
        left = mc.options.keyLeft.isDown
        up = mc.options.keyJump.isDown
        down = mc.options.keyShift.isDown

        unpress()
        if (config.reloadChunks) mc.levelRenderer.allChanged()
    }

    public override fun onDeactivate() {
        super.onDeactivate()
        if (config.reloadChunks) {
            mc.execute { mc.levelRenderer.allChanged() }
        }

        mc.options.cameraType = perspective ?: return

        if (config.staticView) {
            mc.options.fovEffectScale.set(fovScale)
            mc.options.bobView.set(bobView)
        }

        perspective = null
        //isSneaking = false;
    }

    private fun unpress() {
        mc.options.keyUp.isDown = false
        mc.options.keyDown.isDown = false
        mc.options.keyRight.isDown = false
        mc.options.keyLeft.isDown = false
        mc.options.keyJump.isDown = false
        mc.options.keyShift.isDown = false
    }

    @EventHandle
    private fun onOpenScreen(event: ScreenOpenEvent) {
        unpress()

        prevPos.set(pos)
        lastYaw = yaw
        lastPitch = pitch
    }

    @EventHandle
    private fun onTick(event: TickEventPost) {
        Main.LOGGER.warn("Freecam onTick called")
        val cameraEntity = mc.cameraEntity ?: return
        if (cameraEntity.isInWall) cameraEntity.noPhysics = true
        if (perspective == null) return
        if (!perspective!!.isFirstPerson) mc.options.cameraType = CameraType.FIRST_PERSON

        val forward = Vec3.directionFromRotation(0f, yaw)
        val right = Vec3.directionFromRotation(0f, yaw + 90)
        var velX = 0.0
        var velY = 0.0
        var velZ = 0.0

        var s = 0.5
        if (mc.options.keySprint.isDown) s = 1.0

        var a = false
        if (this.forward) {
            velX += forward.x * s * config.speed
            velZ += forward.z * s * config.speed
            a = true
        }
        if (this.backward) {
            velX -= forward.x * s * config.speed
            velZ -= forward.z * s * config.speed
            a = true
        }

        var b = false
        if (this.right) {
            velX += right.x * s * config.speed
            velZ += right.z * s * config.speed
            b = true
        }
        if (this.left) {
            velX -= right.x * s * config.speed
            velZ -= right.z * s * config.speed
            b = true
        }

        if (a && b) {
            val diagonal = 1 / sqrt(2.0)
            velX *= diagonal
            velZ *= diagonal
        }

        if (this.up) {
            velY += s * config.speed
        }
        if (this.down) {
            velY -= s * config.speed
        }

        prevPos.set(pos)
        pos.set(pos.x + velX, pos.y + velY, pos.z + velZ)
    }

    @EventHandle(priority = EventPriority.HIGH)
    fun onKey(event: KeyDownEvent) {
        if (event.keyCode == GLFW.GLFW_KEY_F3) return

        if (onInput(event.keyCode, KeyAction.Press)) event.cancel()
    }

    @EventHandle
    fun onKey(event: KeyUpEvent) {
        if (event.keyCode == GLFW.GLFW_KEY_F3) return

        if (onInput(event.keyCode, KeyAction.Release)) event.cancel()
    }

    @EventHandle(priority = EventPriority.HIGH)
    private fun onMouseClick(event: MouseClickEvent) {
        if (onInput(event.button(), event.action)) event.cancel()
    }

    fun getKey(bind: KeyMapping): Int {
        return (bind as KeyMappingAccessor).`somefrills$getKey`().value
    }

    private fun onInput(key: Int, action: KeyAction): Boolean {
        when (key) {
            getKey(mc.options.keyUp) -> {
                forward = action != KeyAction.Release
                mc.options.keyUp.isDown = false
            }

            getKey(mc.options.keyDown) -> {
                backward = action != KeyAction.Release
                mc.options.keyDown.isDown = false
            }

            getKey(mc.options.keyRight) -> {
                right = action != KeyAction.Release
                mc.options.keyRight.isDown = false
            }

            getKey(mc.options.keyLeft) -> {
                left = action != KeyAction.Release
                mc.options.keyLeft.isDown = false
            }

            getKey(mc.options.keyJump) -> {
                up = action != KeyAction.Release
                mc.options.keyJump.isDown = false
            }

            getKey(mc.options.keyShift) -> {
                down = action != KeyAction.Release
                mc.options.keyShift.isDown = false
            }

            else -> {
                return false
            }
        }

        return true
    }

    @EventHandle(priority = EventPriority.LOW)
    private fun onMouseScroll(event: MouseScrollEvent) {
        if (config.speedScrollSensitivity > 0 && mc.screen == null) {
            config.speed += event.value * 0.25 * (config.speedScrollSensitivity * config.speed)
            if (config.speed < 0.1) config.speed = 0.1

            event.cancel()
        }
    }


    @EventHandle
    private fun onChunkOcclusion(event: ChunkOcclusionEvent) {
        event.cancel()
    }

    @EventHandle
    private fun onGameLeft(event: GameStopEvent) {
        toggle()
    }

    @EventHandle
    private fun onPacketReceive(event: ReceivePacketEvent) {
        val packet = event.packet
        if (packet is ClientboundPlayerCombatKillPacket) {
            val entity = mc.level?.getEntity(packet.playerId()) ?: return
            if (entity === mc.player && config.toggleOnDeath) {
                toggle()
                ChatUtils.info("Toggled off because you died.")
            }
        } else if (packet is ClientboundSetHealthPacket) {
            val player = mc.player ?: return
            if (player.health - packet.health > 0 && config.toggleOnDamage) {
                toggle()
                ChatUtils.info("Toggled off because you took damage.")
            }
        } else if (packet is ClientboundRespawnPacket) {
            toggle()
            ChatUtils.info("Toggled off because you changed dimensions.")
        }
    }

    @EventHandle
    fun onInteractBlock(event: InteractBlockEvent) {
        event.cancel()
    }

    @EventHandle
    fun onInteractEntity(event: InteractEntityEvent) {
        event.cancel()
    }

    @EventHandle(priority = EventPriority.HIGH)
    fun onAttackEntity(event: AttackEntityEvent) {
        event.cancel()
    }

    @EventHandle(priority = EventPriority.HIGH)
    private fun onStartBreakingBlockEvent(event: StartBreakingBlockEvent) {
        event.cancel()
    }

    fun getX(tickDelta: Float): Double {
        return Mth.lerp(tickDelta.toDouble(), prevPos.x, pos.x)
    }

    fun getY(tickDelta: Float): Double {
        return Mth.lerp(tickDelta.toDouble(), prevPos.y, pos.y)
    }

    fun getZ(tickDelta: Float): Double {
        return Mth.lerp(tickDelta.toDouble(), prevPos.z, pos.z)
    }

    fun getYaw(tickDelta: Float): Double {
        return Mth.lerp(tickDelta, lastYaw, yaw).toDouble()
    }

    fun getPitch(tickDelta: Float): Double {
        return Mth.lerp(tickDelta, lastPitch, pitch).toDouble()
    }

    @JvmStatic
    fun shouldRenderHands(): Boolean {
        return config.renderHands
    }

    fun changeLookDirection(deltaX: Double, deltaY: Double) {
        lastYaw = yaw
        lastPitch = pitch

        yaw += deltaX.toFloat()
        pitch += deltaY.toFloat()

        pitch = Mth.clamp(pitch, -90f, 90f)
    }
}